### PR TITLE
Add support for string and time indexed event stats

### PR DESCRIPTION
**Pull request by MatMeredith**
Add support for string and time indexed event stats.   We're going to need these in order to expose per AS latency stats in the S-CSCF.

Tested by https://github.com/Metaswitch/clearwater-fv-test/pull/56

### CODE REVIEW

**Comment by sebrexmetaswitch** [comment]
One thought on this (Adam's reviewing it, but I'm using it for the Sprout AS stats implementation):

Currently, we index the stats by period, and then by string. This means we see stats like:

```
 scopePrevious5SecondPeriod."appserver.domain"                           0                            0                       0                       0                         0
scopePrevious5SecondPeriod."appserver2.domain"                           0                            0                       0                       0                         0
  scopeCurrent5MinutePeriod."appserver.domain"                           0                            0                       0                       0                         0
 scopeCurrent5MinutePeriod."appserver2.domain"                           0                            0                       0                       0                         0
 scopePrevious5MinutePeriod."appserver.domain"                         150                         2500                     200                     100                         2
scopePrevious5MinutePeriod."appserver2.domain"                         300                        10000                     400                     200                         2

```

I'm not sure how easy it would be to do, and if it's hard then it may not be worth it, but would it be nicer to index them by string and then by time period?

(I also don't know if it makes much difference, but I can imagine it would be easier to use the stats if they were indexed by App Server first, then by time period)

File: src/snmp_time_and_string_based_event_table.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  {
+    pthread_rwlock_destroy(&_table_lock);
+
+    //Spin through the maps of 5s and 5m views and delete them.
+    for (auto& kv : _five_second) {  delete kv.second; }
+    for (auto& kv : _five_minute) {  delete kv.second; }
+  }
+private:
+
+  TimeAndStringBasedEventRow* new_row(int indexes) { return NULL;};
```
Comment (johadalin) [inline src/snmp_time_and_string_based_event_table.cpp]: `indicies`, surely :P

**Comment by MatMeredith** [comment]
@sebrexmetaswitch It's a good question.   
- It would be a fair amount of effort to change it I think -- certainly not something we can do this sprint.   On the other hand, once we've shipped an updated MIB, we can't change it around retrospectively.   
- I don't know which way round is better.   My initial instinct is that the other way round would be better, but...  the way it is does mean if what you care about generally is e.g. the previous 5s stats you can say "show me the stats for all app servers for the previous 5s".   And maybe that's actually more useful?   How often are you going to want an overview of all the servers vs wanting to see stats for one specific one?   And if you do want to see stats for a specific one you can of course query it -- you just have to do so separately for each time period.
- Finally (and most importantly!) what we've currently implemented is (I hope!) consistent with all other stats in Clearwater that are indexed by timescope and something else.  See e.g. https://github.com/Metaswitch/clearwater-snmp-handlers/blob/master/mib-generator/CLEARWATER-MIB-COMMON#L1737-L1738 

OK?

File: src/event_statistic_accumulator.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  // structure.
+  stats.count = _count;
+
+  if (_count > 0)
+  {
+    // Get local copies of the values that we are working with atomically.
+    // Note that e.g. _sum is a std::atomic and so assignment is guaranteed
+    // atomic.
+    uint_fast64_t sum = _sum;
+    uint_fast64_t sumsq = _sqsum;
```
Comment (MatMeredith) [inline src/event_statistic_accumulator.cpp]: Reason?  Copied from elsewhere!  Good point though ;-)

File: src/snmp_time_and_string_based_event_table.cpp (hunk 1 above)
Comment (MatMeredith) [inline src/snmp_time_and_string_based_event_table.cpp]: https://english.stackexchange.com/questions/61080/plural-of-index-indexes-or-indices/3126
```
mgm@ip-10-0-0-85:~/clearwater-fv-test/modules/cpp-common$ grep -r indexes *
include/snmp_internal/snmp_table.h:    snmp_free_varbind(_table->indexes_template);
include/snmp_internal/snmp_table.h:    snmp_free_varbind(_table_info->indexes);
include/snmp_internal/snmp_counts_by_other_type_table.h:  RowType* new_row(int indexes) { return NULL; };
include/memcachedstoreview.h:  /// Converts a vector of server indexes into a vector of server names.
include/memcachedstoreview.h:  /// @param ids          - A vector of replica indexes.
src/snmp_continuous_accumulator_by_scope_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_time_and_string_based_event_table.cpp:  TimeAndStringBasedEventRow* new_row(int indexes) { return NULL;};
src/snmp_scalar_by_scope_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_event_accumulator_by_scope_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_continuous_increment_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_counter_by_scope_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_event_accumulator_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_counter_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_success_fail_count_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/memcachedstoreview.cpp:      std::vector<int> server_indexes = ring.get_nodes(ii, replicas);
src/memcachedstoreview.cpp:      for (size_t jj = 0; jj < server_indexes.size(); jj++)
src/memcachedstoreview.cpp:        int idx = server_indexes[jj];
src/snmp_ip_row.cpp:  // Set the IPAddrType and IPAddr as indexes
src/snmp_ip_row.cpp:  // Set the IPAddrType and IPAddr as indexes
src/snmp_continuous_accumulator_table.cpp:  // Map row indexes to the view of the underlying data they should expose
src/snmp_cx_counter_table.cpp:  CxCounterRow* new_row(int indexes) { return NULL;};
mgm@ip-10-0-0-85:~/clearwater-fv-test/modules/cpp-common$ grep -r indicies *
mgm@ip-10-0-0-85:~/clearwater-fv-test/modules/cpp-common$
```

Comment (MatMeredith) [inline src/snmp_time_and_string_based_event_table.cpp]: I will however fix up my incorrect "indices"!

**Comment by MatMeredith** [comment]
@johadalin Thanks.   Re. testing.   As I understand it stats code is deliberately tested in clearwater-fv-test rather than in cpp-common UT because of its dependency on snmp.   I don't plan to do any differently here...

**Pull request merged**